PreferCurlyFors generates illegal code (and it isn't curly either)

- **Version**: 3.0.0
- **Integration**: Coursier launcher
- **Configuration**:
```
version = 3.0.0
rewrite.rules = [ PreferCurlyFors ]
```

## Steps

Given code like this:
```scala
object O {
  for (pos <- "a->1|b->2".split('|'); c = pos.split("->"))
  yield ()
}
```

When I run scalafmt like this:

```bash
scalafmt --config scalafmt.conf S.scala
```

## Problem

Scalafmt formats code like this:
```scala
object O {
  for (
    pos <- "a->1|b->2".split('|')
    c = pos.split("->")
  )
    yield ()
}
```

which the compiler then actually rejects:

```scala
% scalac S.scala
S.scala:4: error: value c is not a member of Array[String]
possible cause: maybe a semicolon is missing before `value c`?
    c = pos.split("->")
    ^
1 error
```

## Expectation

I would like the formatted output to look like this:
```scala
object O {
  for {
    pos <- "a->1|b->2".split('|')
    c = pos.split("->")
  }
    yield ()
}
```

## Notes

this came up over at https://github.com/playframework/twirl/pull/437

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala
@@ -17,8 +17,10 @@ object PreferCurlyFors extends Rewrite with FormatTokensRewrite.RuleFactory {
   override def create(ftoks: FormatTokens): FormatTokensRewrite.Rule =
     new PreferCurlyFors(ftoks)
 
-  private def hasMultipleGenerators(enums: Seq[Enumerator]): Boolean =
-    enums.count(_.is[Enumerator.Generator]) > 1
+  @inline
+  private def hasMultipleNonGuardEnums(enums: Seq[Enumerator]): Boolean =
+    // first one is never a guard
+    enums.view.drop(1).exists(!_.is[Enumerator.Guard])
 
   case class Settings(
       removeTrailingSemicolonsOnly: Boolean = false
@@ -64,8 +66,8 @@ private class PreferCurlyFors(ftoks: FormatTokens)
       case x: Token.LeftParen
           if ftoks.prevNonComment(ft).left.is[Token.KwFor] =>
         val ok = ft.meta.rightOwner match {
-          case t: Term.For => hasMultipleGenerators(t.enums)
-          case t: Term.ForYield => hasMultipleGenerators(t.enums)
+          case t: Term.For => hasMultipleNonGuardEnums(t.enums)
+          case t: Term.ForYield => hasMultipleNonGuardEnums(t.enums)
           case _ => false
         }
         if (ok)

--- a/scalafmt-tests/src/test/resources/rewrite/PreferCurlyFors.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/PreferCurlyFors.stat
@@ -128,3 +128,29 @@ object a {
     println(a + b)
   }
 }
+<<< #2695 generator and val
+object O {
+  for (pos <- "a->1|b->2".split('|'); c = pos.split("->"))
+  yield ()
+}
+>>>
+object O {
+  for (pos <- "a->1|b->2".split('|') c = pos.split("->"))
+Idempotency violated
+  for (
+    pos <- "a->1|b->2".split('|')
+    c = pos.split("->")
+  )
+    yield ()
+}
+<<< #2695 generator and multiple guards
+object O {
+  for (a <- b; if c; if d) yield ()
+}
+>>>
+object O {
+  for (
+    a <- b if c
+    if d
+  ) yield ()
+}

--- a/scalafmt-tests/src/test/resources/rewrite/PreferCurlyFors.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/PreferCurlyFors.stat
@@ -135,13 +135,10 @@ object O {
 }
 >>>
 object O {
-  for (pos <- "a->1|b->2".split('|') c = pos.split("->"))
-Idempotency violated
-  for (
+  for {
     pos <- "a->1|b->2".split('|')
     c = pos.split("->")
-  )
-    yield ()
+  } yield ()
 }
 <<< #2695 generator and multiple guards
 object O {

